olp-cpp-sdk-authentication/src/AuthenticationClient.cpp

```diff
@@ -477,66 +491,99 @@ client::CancellationToken AuthenticationClient::Impl::SignInClient(
+
+      if (get_timestamp) {
+        get_timestamp(auth_response.headers);
       }
-    }
+      const auto status = auth_response.status;
 
-    const auto url = settings_.token_endpoint_url + kOauthEndpoint;
+      if (status < 0) {
+        auto error_message = auth_response.response.str();
 
-    client::OlpClient::ParametersType headers = {
-        {http::kAuthorizationHeader,
-         GenerateHeader(credentials, url, timestamp)}};
+        if (status == static_cast<int>(olp::http::ErrorCode::CANCELLED_ERROR)) {
+          return {{status, error_message}};
+        }
 
-    auto auth_response = client.CallApi(
-        kOauthEndpoint, "POST", {}, std::move(headers), {},
-        GenerateClientBody(properties), kApplicationJson, context);
+        auto cached_response = client_token_cache_->locked(
+            [&](utils::LruCache<std::string, SignInResult>& c)
+                -> boost::optional<SignInResult> {
+              auto it = c.Find(credentials.GetKey());
+              return it != c.end() ? boost::make_optional(it->value())
+                                   : boost::none;
+            });
 
-    const auto status = auth_response.status;
+        if (!cached_response) {
+          return {{status, error_message}};
+        }
 
-    if (status < 0) {
-      auto error_message = auth_response.response.str();
+        return *cached_response;
+      }
 
-      if (status == static_cast<int>(olp::http::ErrorCode::CANCELLED_ERROR)) {
-        return {{status, error_message}};
+      auto document = std::make_shared<rapidjson::Document>();
+      rapidjson::IStreamWrapper stream(auth_response.response);
+      document->ParseStream(stream);
+      auto resp_impl = std::make_shared<SignInResultImpl>(
+          status, olp::http::HttpErrorToString(status), document);
+      SignInResult response(resp_impl);
+
+      if (status == http::HttpStatusCode::OK) {
+        // Cache the response
+        client_token_cache_->locked(
+            [&](utils::LruCache<std::string, SignInResult>& c) {
+              return c.InsertOrAssign(credentials.GetKey(), response);
+            });
       }
 
-      auto cached_response = client_token_cache_->locked(
-          [&](utils::LruCache<std::string, SignInResult>& c)
-              -> boost::optional<SignInResult> {
-            auto it = c.Find(credentials.GetKey());
-            return it != c.end() ? boost::make_optional(it->value())
-                                 : boost::none;
-          });
+      return response;
+    };
+    std::time_t timestamp = 0;
+    std::time_t timestamp_from_header = 0;
+    ProcessReturnHeaders get_timestamp_from_headers = nullptr;
 
-      if (!cached_response) {
-        return {{status, error_message}};
+    if (!settings_.use_system_time) {
+      auto server_time = GetTimeFromServer(context, client);
+      if (server_time.IsSuccessful()) {
+        timestamp = server_time.GetResult();
       }
```

Comment:
Just curious, are we ok with failed  GetTimeFromServer()?
I think, we should have some log message and return the error.